core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -705,74 +727,77 @@ public SourceDistributionTaskSource(
         }
 
         @Override
-        public List<TaskDescriptor> getMoreTasks()
+        public ListenableFuture<List<TaskDescriptor>> getMoreTasks()
         {
             if (finished || closed) {
-                return ImmutableList.of();
+                return immediateFuture(ImmutableList.of());
             }
 
-            List<TaskDescriptor> result = new ArrayList<>();
-
-            boolean splitSourceFinished = false;
-            while (result.isEmpty()) {
-                ListenableFuture<SplitBatch> splitBatchFuture = splitSource.getNextBatch(NOT_PARTITIONED, Lifespan.taskWide(), splitBatchSize);
+            List<TaskDescriptor> result = getAllReadyTasks();
+            if (!result.isEmpty()) {
+                return immediateFuture(result);
+            }
 
-                long start = System.nanoTime();
-                addSuccessCallback(splitBatchFuture, () -> getSplitTimeRecorder.accept(start));
+            ListenableFuture<SplitBatch> splitBatchFuture = splitSource.getNextBatch(NOT_PARTITIONED, Lifespan.taskWide(), splitBatchSize);
 
-                SplitBatch splitBatch = getFutureValue(splitBatchFuture);
-                List<Split> splits = splitBatch.getSplits();
+            long start = System.nanoTime();
+            addSuccessCallback(splitBatchFuture, () -> getSplitTimeRecorder.accept(start));
 
-                for (Split split : splits) {
-                    if (split.isRemotelyAccessible()) {
-                        remotelyAccessibleSplitBuffer.add(split);
-                    }
-                    else {
-                        List<HostAddress> addresses = split.getAddresses();
-                        checkArgument(!addresses.isEmpty(), "split is not remotely accessible but the list of addresses is empty");
-                        for (HostAddress hostAddress : addresses) {
-                            locallyAccessibleSplitBuffer.computeIfAbsent(hostAddress, key -> newIdentityHashSet()).add(split);
+            return Futures.transform(
+                    splitBatchFuture,
+                    splitBatch -> {
+                        for (Split split : splitBatch.getSplits()) {
+                            if (split.isRemotelyAccessible()) {
+                                remotelyAccessibleSplitBuffer.add(split);
```

Comment:
This isn't thread safe. We should run the callback under a lock, mark mutable fields as `@GuardedBy` and enforce thread safety for

```
        private final Set<Split> remotelyAccessibleSplitBuffer = newIdentityHashSet();
        private final Map<HostAddress, Set<Split>> locallyAccessibleSplitBuffer = new HashMap<>();

        private int currentPartitionId;
        private boolean finished;
        private boolean closed;
```